File: scripts/pipeline_fixture.py
#!/usr/bin/env python3
"""Run the deterministic scrape-to-query pipeline fixture.

The runner uses only Python's standard library for local test doubles. It
starts the real Ohara scraper, cleaning, indexer, and retrieval binaries, then
checks their JSON/file boundaries with no external network or model download.
"""

from __future__ import annotations

import json
import os
from pathlib import Path
import signal
import socket
import subprocess
import sys
import tempfile
import threading
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.parse import urlsplit
from urllib.request import Request, urlopen


ROOT = Path(__file__).resolve().parents[1]
POLL_INTERVAL_SECONDS = 0.1
PROCESS_START_TIMEOUT_SECONDS = 15.0
PIPELINE_TIMEOUT_SECONDS = 30.0


def free_port() -> int:
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as listener:
        listener.bind(("127.0.0.1", 0))
        return int(listener.getsockname()[1])


class QuietHandler(BaseHTTPRequestHandler):
    def log_message(self, _format: str, *_args: Any) -> None:
        return

    def send_json(self, status: int, payload: Any) -> None:
        body = json.dumps(payload).encode()
        self.send_response(status)
        self.send_header("content-type", "application/json")
        self.send_header("content-length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)


class FixtureHandler(QuietHandler):
    def do_GET(self) -> None:  # noqa: N802
        if self.path.split("?", 1)[0] == "/news":
            article_url = (
                f"http://127.0.0.1:{self.server.server_port}/article.html"
            )
            body = f"""<?xml version="1.0" encoding="UTF-8"?>
<rss version="2.0"><channel><title>Ohara fixture feed</title>
<item><title>Ohara fixture article</title><link>{article_url}</link></item>
</channel></rss>""".encode()
            self.send_response(200)
            self.send_header("content-type", "application/rss+xml")
            self.send_header("content-length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)
            return

        if self.path == "/article.html":
            body = """<!doctype html>
<html><head><title>Ohara fixture article</title></head>
<body><article>
<h1>Ohara fixture article</h1>
<p>Ohara is a private local knowledge base that turns web topics into searchable evidence.</p>
<p>The scraper fetches source pages, cleaning extracts the article, and the indexer creates searchable chunks.</p>
<p>Retrieval finds the relevant evidence and returns a local grounded answer with citations.</p>
</article></body></html>""".encode()
            self.send_response(200)
            self.send_header("content-type", "text/html")
            self.send_header("content-length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)
            return

        self.send_error(404)


class QdrantHandler(QuietHandler):
    def do_GET(self) -> None:  # noqa: N802
        if urlsplit(self.path).path == "/collections":
            self.send_json(200, {"result": {"collections": []}, "status": "ok"})
            return
        self.send_error(404)

    def do_PUT(self) -> None:  # noqa: N802
        path = urlsplit(self.path).path
        if path.endswith("/points"):
            payload = read_json(self)
            points = payload.get("points", [])
            if not isinstance(points, list) or any(
                not isinstance(point, dict)
                or not isinstance(point.get("vector"), list)
                or len(point["vector"]) != 384
                for point in points
            ):
                self.send_json(400, {"status": "invalid fixture vector"})
                return
            self.server.points.extend(points)
            self.send_json(200, {"result": True, "status": "ok"})
            return
        if path.startswith("/collections/"):
            self.send_json(200, {"result": True, "status": "ok"})
            return
        self.send_error(404)

    def do_POST(self) -> None:  # noqa: N802
        if urlsplit(self.path).path.endswith("/points/search"):
            payload = read_json(self)
            vector = payload.get("vector")
            if not isinstance(vector, list) or len(vector) != 384:
                self.send_json(400, {"status": "invalid fixture query vector"})
                return
            limit = int(payload.get("limit", len(self.server.points)))
            points = [
                {
                    "id": point.get("id"),
                    "score": 1.0,
                    "payload": point.get("payload", {}),
                }
                for point in self.server.points[:limit]
            ]
            self.send_json(200, {"result": points, "status": "ok"})
            return
        self.send_error(404)


class OllamaHandler(QuietHandler):
    def do_GET(self) -> None:  # noqa: N802
        if self.path == "/api/tags":
            self.send_json(200, {"models": [{"name": "fixture"}]})
            return
        self.send_error(404)

    def do_POST(self) -> None:  # noqa: N802
        if self.path == "/api/generate":
            read_json(self)
            self.send_json(
                200,
                {
                    "response": (
                        "Ohara is a private local knowledge base built from local artifacts."
                    )
                },
            )
            return
        self.send_error(404)


class ProviderServer(ThreadingHTTPServer):
    allow_reuse_address = True

    def __init__(self, handler: type[BaseHTTPRequestHandler]) -> None:
        super().__init__(("127.0.0.1", 0), handler)
        self.points: list[dict[str, Any]] = []


def read_json(handler: BaseHTTPRequestHandler) -> dict[str, Any]:
    length = int(handler.headers.get("content-length", "0"))
    value = json.loads(handler.rfile.read(length))
    if not isinstance(value, dict):
        raise ValueError("expected a JSON object")
    return value


def start_provider(handler: type[BaseHTTPRequestHandler]) -> ProviderServer:
    server = ProviderServer(handler)
    threading.Thread(target=server.serve_forever, daemon=True).start()
    return server


def request_json(
    method: str,
    url: str,
    payload: dict[str, Any] | None = None,
    headers: dict[str, str] | None = None,
) -> tuple[int, Any]:
    body = None if payload is None else json.dumps(payload).encode()
    request_headers = {
        "accept": "application/json",
        "content-type": "application/json",
    }
    if headers:
        request_headers.update(headers)
    request = Request(
        url,
        data=body,
        headers=request_headers,
        method=method,
    )
    try:
        with urlopen(request, timeout=2) as response:
            response_body = response.read()
            return response.status, json.loads(response_body) if response_body else None
    except HTTPError as error:
        response_body = error.read()
        try:
            payload_value = json.loads(response_body) if response_body else None
        except json.JSONDecodeError:
            payload_value = response_body.decode(errors="replace")
        return error.code, payload_value


def wait_for_http(url: str, expected_status: int | None = None) -> None:
    deadline = time.monotonic() + PROCESS_START_TIMEOUT_SECONDS
    while time.monotonic() < deadline:
        try:
            status, _payload = request_json("GET", url)
            if expected_status is None or status == expected_status:
                return
        except (OSError, URLError, ValueError):
            pass
        time.sleep(POLL_INTERVAL_SECONDS)
    raise RuntimeError(f"timed out waiting for {url}")


def wait_for_file(path: Path, processes: list["ManagedProcess"]) -> None:
    deadline = time.monotonic() + PIPELINE_TIMEOUT_SECONDS
    while time.monotonic() < deadline:
        if path.is_file():
            return
        for process in processes:
            if process.process.poll() is not None:
                raise RuntimeError(
                    f"{process.name} exited with status {process.process.returncode}"
                )
        time.sleep(POLL_INTERVAL_SECONDS)
    raise RuntimeError(f"timed out waiting for {path}")


class ManagedProcess:
    def __init__(self, name: str, binary: str, environment: dict[str, str], log_dir: Path):
        self.name = name
        self.log_path = log_dir / f"{name}.log"
        self.log_file = self.log_path.open("w", encoding="utf-8")
        self.process = subprocess.Popen(
            [str(ROOT / "target" / "debug" / binary)],
            cwd=ROOT,
            env=environment,
            stdout=self.log_file,
            stderr=subprocess.STDOUT,
            text=True,
        )

    def close(self) -> None:
        self.log_file.close()


def check(condition: bool, message: str) -> None:
    if not condition:
        raise RuntimeError(message)


def run_pipeline() -> None:
    binaries = {
        "scraper": "ohara-scraper",
        "cleaning": "ohara-cleaning",
        "indexer": "ohara-indexer",
        "retrieval": "ohara-retrieval",
    }
    with tempfile.TemporaryDirectory(prefix="ohara-pipeline-") as temporary:
        data_dir = Path(temporary) / "data"
        log_dir = Path(temporary) / "logs"
        data_dir.mkdir()
        log_dir.mkdir()
        fixture_server = start_provider(FixtureHandler)
        qdrant_server = start_provider(QdrantHandler)
        ollama_server = start_provider(OllamaHandler)
        processes: list[ManagedProcess] = []
        try:
            scraper_port = free_port()
            retrieval_port = free_port()
            base_environment = os.environ.copy()
            base_environment.update(
                {
                    "OHARA_DATA_DIR": str(data_dir),
                    "OHARA_QDRANT_URL": f"http://127.0.0.1:{qdrant_server.server_port}",
                    "OHARA_LLM_URL": f"http://127.0.0.1:{ollama_server.server_port}",
                    "OHARA_LLM_MODEL": "fixture",
                    "OHARA_EMBEDDING_MODE": "deterministic",
                    "OHARA_PROCESS_AUTH_TOKEN": "fixture-process-token",
                }
            )

            scraper_environment = base_environment | {
                "OHARA_SCRAPER_BIND": f"127.0.0.1:{scraper_port}",
                "OHARA_SCRAPER_SEARCH_URL": (
                    f"http://127.0.0.1:{fixture_server.server_port}/news"
                ),
            }
            retrieval_environment = base_environment | {
                "OHARA_RETRIEVAL_BIND": f"127.0.0.1:{retrieval_port}",
                "OHARA_SCRAPER_URL": f"http://127.0.0.1:{scraper_port}",
            }
            processes.extend(
                [
                    ManagedProcess("scraper", binaries["scraper"], scraper_environment, log_dir),
                    ManagedProcess("cleaning", binaries["cleaning"], base_environment, log_dir),
                    ManagedProcess("indexer", binaries["indexer"], base_environment, log_dir),
                    ManagedProcess(
                        "retrieval", binaries["retrieval"], retrieval_environment, log_dir
                    ),
                ]
            )

            wait_for_http(f"http://127.0.0.1:{scraper_port}/health", expected_status=204)
            wait_for_http(f"http://127.0.0.1:{retrieval_port}/api/health")
            status, auth_payload = request_json(
                "POST",
                f"http://127.0.0.1:{scraper_port}/scrape",
                {"topic": "deterministic fixture topic", "limit": 1},
            )
            check(
                status == 401,
                f"unauthenticated scrape returned HTTP {status}: {auth_payload}",
            )
            status, scrape_payload = request_json(
                "POST",
                f"http://127.0.0.1:{retrieval_port}/api/topics/scrape",
                {"topic": "deterministic fixture topic", "limit": 1},
            )
            check(status == 200, f"scrape returned HTTP {status}: {scrape_payload}")
            check(isinstance(scrape_payload, dict), "scrape did not return an object")
            check(scrape_payload.get("discovered") == 1, "fixture result was not discovered")
            check(scrape_payload.get("enqueued") == 1, "fixture result was not enqueued")
            documents = scrape_payload.get("documents")
            check(isinstance(documents, list) and len(documents) == 1, "missing fixture document")
            document_id = documents[0]["documentId"]

            raw_artifact = data_dir / "raw" / f"{document_id}.json"
            clean_artifact = data_dir / "clean" / f"{document_id}.json"
            indexed_artifact = data_dir / "indexed" / f"{document_id}.json"
            wait_for_file(raw_artifact, processes)
            wait_for_file(clean_artifact, processes)
            wait_for_file(indexed_artifact, processes)

            clean_payload = json.loads(clean_artifact.read_text(encoding="utf-8"))
            indexed_payload = json.loads(indexed_artifact.read_text(encoding="utf-8"))
            check(
                "Ohara is a private local knowledge base" in clean_payload["markdown"],
                "clean artifact did not contain the fixture article",
            )
            check(
                indexed_payload["chunks"]
                and indexed_payload["chunks"][0]["document_id"] == document_id,
                "indexed artifact did not contain the fixture chunk",
            )

            check(not (data_dir / "inbox/cleaning" / f"{document_id}.json").exists(), "cleaning inbox item remained")
            check(not (data_dir / "inbox/indexer" / f"{document_id}.json").exists(), "indexer inbox item remained")
            check((data_dir / "inbox/graph" / f"{document_id}.json").is_file(), "graph handoff was not published")

            status, query_payload = request_json(
                "POST",
                f"http://127.0.0.1:{retrieval_port}/api/query",
                {"query": "What is Ohara?", "top_k": 1},
            )
            check(status == 200, f"query returned HTTP {status}: {query_payload}")
            check(isinstance(query_payload, dict), "query did not return an object")
            check(query_payload.get("grounding") == "grounded", "query was not grounded")
            check(
                isinstance(query_payload.get("answer"), str)
                and query_payload["answer"].strip(),
                "query answer was empty",
            )
            chunks = query_payload.get("chunks")
            citations = query_payload.get("citations")
            check(isinstance(chunks, list) and chunks, "query returned no chunks")
            check(isinstance(citations, list) and citations, "query returned no citations")
            check(citations[0] == chunks[0]["chunkId"], "citation did not reference evidence")
            print(f"pipeline fixture passed: {document_id}")
        finally:
            for process in reversed(processes):
                if process.process.poll() is None:
                    process.process.send_signal(signal.SIGTERM)
            for process in reversed(processes):
                try:
                    process.process.wait(timeout=5)
                except subprocess.TimeoutExpired:
                    process.process.kill()
                    process.process.wait(timeout=5)
                process.close()
            fixture_server.shutdown()
            qdrant_server.shutdown()
            ollama_server.shutdown()
            if any(process.process.returncode not in (0, -signal.SIGTERM) for process in processes):
                for process in processes:
                    output = process.log_path.read_text(encoding="utf-8")
                    if output:
                        print(f"--- {process.name} log ---\n{output}", file=sys.stderr)


def main() -> int:
    try:
        run_pipeline()
    except (OSError, RuntimeError, ValueError) as error:
        print(f"pipeline fixture failed: {error}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/src/pipeline_fixture.rs
//! Real process-seam fixture for scrape-to-query verification.

use crate::artifacts::{read_json, temporary_directory};
use crate::http::{client, free_port, request_json, wait_for_http};
use crate::process::{ManagedProcess, environment, with_values};
use crate::providers::{
    FixtureServer, OllamaServer, QdrantServer, start_fixture_server, start_ollama_server,
    start_qdrant_server,
};
use crate::{Error, Result, check, repository_root};
use reqwest::Client;
use serde_json::{Value, json};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::time::Duration;

const PROCESS_START_TIMEOUT: Duration = Duration::from_secs(15);
const PIPELINE_TIMEOUT: Duration = Duration::from_secs(30);

/// Execute the deterministic scrape-to-query fixture.
///
/// # Errors
///
/// Returns an error when a fixture provider, Ohara process, artifact, or
/// response contract fails validation.
pub async fn run() -> Result<()> {
    let mut setup = PipelineSetup::start().await?;
    match setup.execute().await {
        Ok(()) => setup.cleanup(true).await,
        Err(error) => {
            if let Err(cleanup_error) = setup.cleanup(false).await {
                eprintln!("pipeline fixture cleanup failed: {cleanup_error}");
            }
            Err(error)
        }
    }
}

struct PipelineSetup {
    root: PathBuf,
    temporary: PathBuf,
    data_dir: PathBuf,
    log_dir: PathBuf,
    http_client: Client,
    fixture_server: FixtureServer,
    qdrant_server: QdrantServer,
    ollama_server: OllamaServer,
    scraper_port: u16,
    retrieval_port: u16,
    scraper_environment: HashMap<String, String>,
    base_environment: HashMap<String, String>,
    retrieval_environment: HashMap<String, String>,
    processes: Vec<ManagedProcess>,
}

impl PipelineSetup {
    async fn start() -> Result<Self> {
        let root = repository_root()?;
        let (temporary, data_dir, log_dir) = create_workspace()?;
        let http_client = client()?;
        let fixture_server = start_fixture_server().await?;
        let search_mode = pipeline_search_mode()?;
        let qdrant_server = start_qdrant_server(Some(&search_mode)).await?;
        let ollama_server = start_ollama_server().await?;
        let scraper_port = free_port()?;
        let retrieval_port = free_port()?;
        let scraper_config = write_scraper_config(&temporary, scraper_port, &fixture_server)?;
        let base_environment = base_environment(
            &data_dir,
            qdrant_server.port(),
            ollama_server.port(),
            &search_mode,
        );
        let scraper_environment = with_values(
            &base_environment,
            [("OHARA_SCRAPER_CONFIG".to_owned(), scraper_config)],
        );
        let retrieval_environment = with_values(
            &base_environment,
            [
                (
                    "OHARA_RETRIEVAL_BIND".to_owned(),
                    format!("127.0.0.1:{retrieval_port}"),
                ),
                (
                    "OHARA_SCRAPER_URL".to_owned(),
                    format!("http://127.0.0.1:{scraper_port}"),
                ),
            ],
        );
        Ok(Self {
            root,
            temporary,
            data_dir,
            log_dir,
            http_client,
            fixture_server,
            qdrant_server,
            ollama_server,
            scraper_port,
            retrieval_port,
            scraper_environment,
            base_environment,
            retrieval_environment,
            processes: Vec::new(),
        })
    }

    async fn execute(&mut self) -> Result<()> {
        self.start_processes()?;
        self.wait_for_services().await?;
        let document_id = self.scrape().await?;
        self.assert_artifacts(&document_id).await?;
        self.assert_grounded_query().await?;
        self.assert_model_contracts().await?;
        println!("pipeline fixture passed: {document_id}");
        Ok(())
    }

    fn start_processes(&mut self) -> Result<()> {
        for (name, binary, environment) in [
            ("scraper", "ohara-scraper", &self.scraper_environment),
            ("cleaning", "ohara-cleaning", &self.base_environment),
            ("indexer", "ohara-indexer", &self.base_environment),
            ("retrieval", "ohara-retrieval", &self.retrieval_environment),
        ] {
            self.processes.push(ManagedProcess::start(
                &self.root,
                name,
                binary,
                environment,
                &self.log_dir,
            )?);
        }
        Ok(())
    }

    async fn wait_for_services(&mut self) -> Result<()> {
        wait_for_http(
            &self.http_client,
            &format!("http://127.0.0.1:{}/health", self.scraper_port),
            Some(204),
            PROCESS_START_TIMEOUT,
        )
        .await?;
        wait_for_http(
            &self.http_client,
            &format!("http://127.0.0.1:{}/api/health", self.retrieval_port),
            None,
            PROCESS_START_TIMEOUT,
        )
        .await
    }

    async fn scrape(&self) -> Result<String> {
        let scraper_url = format!("http://127.0.0.1:{}", self.scraper_port);
        let (status, auth_payload) = request_json(
            &self.http_client,
            "POST",
            &format!("{scraper_url}/scrape"),
            Some(&json!({"topic": "deterministic fixture topic", "limit": 1})),
        )
        .await?;
        check(
            status == 401,
            format!("unauthenticated scrape returned HTTP {status}: {auth_payload}"),
        )?;
        let (status, payload) = request_json(
            &self.http_client,
            "POST",
            &format!("{}/api/topics/scrape", self.retrieval_url()),
            Some(&json!({"topic": "deterministic fixture topic", "limit": 1})),
        )
        .await?;
        check(
            status == 200,
            format!("scrape returned HTTP {status}: {payload}"),
        )?;
        check(payload.is_object(), "scrape did not return an object")?;
        check(
            payload.get("discovered") == Some(&json!(1)),
            "fixture was not discovered",
        )?;
        check(
            payload.get("enqueued") == Some(&json!(1)),
            "fixture was not enqueued",
        )?;
        let documents = payload
            .get("documents")
            .and_then(Value::as_array)
            .ok_or_else(|| Error::Message("missing fixture documents".to_owned()))?;
        check(documents.len() == 1, "missing fixture document")?;
        documents[0]
            .get("documentId")
            .and_then(Value::as_str)
            .map(str::to_owned)
            .ok_or_else(|| Error::Message("fixture document has no documentId".to_owned()))
    }

    async fn assert_artifacts(&mut self, document_id: &str) -> Result<()> {
        let artifact_paths = [
            self.data_dir
                .join("raw")
                .join(format!("{document_id}.json")),
            self.data_dir
                .join("clean")
                .join(format!("{document_id}.json")),
            self.data_dir
                .join("indexed")
                .join(format!("{document_id}.json")),
        ];
        for path in &artifact_paths {
            wait_for_file(path, &mut self.processes).await?;
        }
        let clean_payload = read_json(&artifact_paths[1])?;
        let indexed_payload = read_json(&artifact_paths[2])?;
        check(
            clean_payload
                .get("markdown")
                .and_then(Value::as_str)
                .is_some_and(|value| value.contains("Ohara is a private local knowledge base")),
            "clean artifact did not contain the fixture article",
        )?;
        check(
            indexed_payload
                .get("chunks")
                .and_then(Value::as_array)
                .and_then(|chunks| chunks.first())
                .and_then(|chunk| chunk.get("document_id"))
                .and_then(Value::as_str)
                == Some(document_id),
            "indexed artifact did not contain the fixture chunk",
        )?;
        self.assert_handoffs(document_id).await
    }

    async fn assert_handoffs(&mut self, document_id: &str) -> Result<()> {
        for process in ["cleaning", "indexer"] {
            let path = self
                .data_dir
                .join(format!("inbox/{process}"))
                .join(format!("{document_id}.json"));
            wait_for_absence(&path, &mut self.processes).await?;
        }
        check(
            self.data_dir
                .join("inbox/graph")
                .join(format!("{document_id}.json"))
                .is_file(),
            "graph handoff was not published",
        )
    }

    async fn assert_grounded_query(&self) -> Result<()> {
        let (status, payload) = self.query().await?;
        check(
            status == 200,
            format!("query returned HTTP {status}: {payload}"),
        )?;
        check(
            payload.get("grounding") == Some(&json!("grounded")),
            "query was not grounded",
        )?;
        check(
            payload
                .get("answer")
                .and_then(Value::as_str)
                .is_some_and(|answer| !answer.trim().is_empty()),
            "query answer was empty",
        )?;
        let chunks = payload
            .get("chunks")
            .and_then(Value::as_array)
            .ok_or_else(|| Error::Message("query returned no chunks".to_owned()))?;
        check(!chunks.is_empty(), "query returned no chunks")?;
        let citations = payload
            .get("citations")
            .and_then(Value::as_array)
            .ok_or_else(|| Error::Message("query returned no citations".to_owned()))?;
        check(!citations.is_empty(), "query returned no citations")?;
        let signals = payload
            .get("signals")
            .and_then(Value::as_object)
            .ok_or_else(|| Error::Message("query returned no signal metadata".to_owned()))?;
        check(
            signals.get("qdrant") == Some(&json!(true)),
            "qdrant signal was unavailable",
        )?;
        check(
            signals.get("fullText") == Some(&json!(true)),
            "full-text signal was unavailable",
        )?;
        check(
            signals.get("graphPath").is_some_and(Value::is_boolean),
            "graph signal metadata was invalid",
        )?;
        check(
            citations.first() == chunks.first().and_then(|chunk| chunk.get("chunkId")),
            "citation did not reference evidence",
        )
    }

    async fn assert_model_contracts(&mut self) -> Result<()> {
        for (mode, case_name, availability) in [
            ("empty", "empty-model", "available"),
            ("unavailable", "unavailable-model", "unavailable"),
            ("malformed", "malformed-model", "unavailable"),
        ] {
            self.ollama_server.set_response_mode(mode).await?;
            let (status, payload) = self.query().await?;
            assert_contract(status, &payload, case_name, availability)?;
        }
        Ok(())
    }

    async fn query(&self) -> Result<(u16, Value)> {
        request_json(
            &self.http_client,
            "POST",
            &format!("{}/api/query", self.retrieval_url()),
            Some(&json!({"query": "What is Ohara?", "top_k": 1})),
        )
        .await
    }

    fn retrieval_url(&self) -> String {
        format!("http://127.0.0.1:{}", self.retrieval_port)
    }

    async fn cleanup(&mut self, passed: bool) -> Result<()> {
        stop_processes(&mut self.processes).await;
        if passed {
            verify_drain_logs(&self.processes)?;
        } else {
            print_process_logs(&self.processes)?;
        }
        self.fixture_server.shutdown().await?;
        self.qdrant_server.shutdown().await?;
        self.ollama_server.shutdown().await?;
        let _ = std::fs::remove_dir_all(&self.temporary);
        Ok(())
    }
}

fn assert_contract(
    status: u16,
    payload: &Value,
    case_name: &str,
    availability: &str,
) -> Result<()> {
    check(
        status == 200,
        format!("{case_name} query returned HTTP {status}: {payload}"),
    )?;
    check(
        payload.get("availability") == Some(&json!(availability))
            && payload.get("grounding") == Some(&json!("ungrounded"))
            && payload.get("answer") == Some(&Value::Null)
            && payload.get("citations") == Some(&json!([])),
        format!("{case_name} contract was invalid: {payload}"),
    )
}

fn create_workspace() -> Result<(PathBuf, PathBuf, PathBuf)> {
    let temporary = temporary_directory("ohara-pipeline")?;
    let data_dir = temporary.join("data");
    let log_dir = temporary.join("logs");
    std::fs::create_dir_all(&data_dir)
        .map_err(|source| Error::io("create pipeline data", source))?;
    std::fs::create_dir_all(&log_dir)
        .map_err(|source| Error::io("create pipeline logs", source))?;
    Ok((temporary, data_dir, log_dir))
}

fn pipeline_search_mode() -> Result<String> {
    let mode =
        std::env::var("OHARA_PIPELINE_QDRANT_SEARCH_MODE").unwrap_or_else(|_| "exact".to_owned());
    check(
        matches!(mode.as_str(), "exact" | "hnsw"),
        "pipeline Qdrant search mode must be exact or hnsw",
    )?;
    Ok(mode)
}

fn write_scraper_config(
    temporary: &Path,
    scraper_port: u16,
    fixture_server: &FixtureServer,
) -> Result<String> {
    let path = temporary.join("scraper-config.yaml");
    let config = format!(
        "bind: 127.0.0.1:{scraper_port}\nsearch:\n  provider: rss\n  url: http://127.0.0.1:{}/news\nfetch:\n  kind: http\n",
        fixture_server.port()
    );
    std::fs::write(&path, config)
        .map_err(|source| Error::io("write pipeline scraper config", source))?;
    path.to_str()
        .map(str::to_owned)
        .ok_or_else(|| Error::Message("pipeline scraper config path is not valid UTF-8".to_owned()))
}

fn base_environment(
    data_dir: &Path,
    qdrant_port: u16,
    ollama_port: u16,
    search_mode: &str,
) -> HashMap<String, String> {
    environment([
        ("OHARA_DATA_DIR".to_owned(), data_dir.display().to_string()),
        (
            "OHARA_QDRANT_URL".to_owned(),
            format!("http://127.0.0.1:{qdrant_port}"),
        ),
        (
            "OHARA_FALKORDB_URL".to_owned(),
            "redis://127.0.0.1:1".to_owned(),
        ),
        (
            "OHARA_LLM_URL".to_owned(),
            format!("http://127.0.0.1:{ollama_port}"),
        ),
        ("OHARA_LLM_MODEL".to_owned(), "fixture".to_owned()),
        (
            "OHARA_EMBEDDING_MODE".to_owned(),
            "deterministic".to_owned(),
        ),
        (
            "OHARA_QDRANT_SEARCH_MODE".to_owned(),
            search_mode.to_owned(),
        ),
        (
            "OHARA_QDRANT_COLLECTION".to_owned(),
            "ohara_chunks".to_owned(),
        ),
        (
            "OHARA_PROCESS_AUTH_TOKEN".to_owned(),
            "fixture-process-token".to_owned(),
        ),
    ])
}

async fn stop_processes(processes: &mut [ManagedProcess]) {
    for process in processes.iter_mut().rev() {
        if let Err(error) = process.stop().await {
            eprintln!("{} cleanup failed: {error}", process.name);
        }
    }
}

fn verify_drain_logs(processes: &[ManagedProcess]) -> Result<()> {
    for process in processes {
        if matches!(process.name.as_str(), "cleaning" | "indexer") {
            check(
                process.log()?.contains("drain complete"),
                format!("{} did not complete its graceful drain", process.name),
            )?;
        }
    }
    Ok(())
}

async fn wait_for_file(path: &Path, processes: &mut [ManagedProcess]) -> Result<()> {
    let deadline = tokio::time::Instant::now() + PIPELINE_TIMEOUT;
    while tokio::time::Instant::now() < deadline {
        if path.is_file() {
            return Ok(());
        }
        ensure_processes_running(processes)?;
        tokio::time::sleep(Duration::from_millis(100)).await;
    }
    Err(Error::Message(format!(
        "timed out waiting for {}",
        path.display()
    )))
}

async fn wait_for_absence(path: &Path, processes: &mut [ManagedProcess]) -> Result<()> {
    let deadline = tokio::time::Instant::now() + PIPELINE_TIMEOUT;
    while tokio::time::Instant::now() < deadline {
        if !path.exists() {
            return Ok(());
        }
        ensure_processes_running(processes)?;
        tokio::time::sleep(Duration::from_millis(100)).await;
    }
    Err(Error::Message(format!(
        "timed out waiting for {} to be removed",
        path.display()
    )))
}

fn ensure_processes_running(processes: &mut [ManagedProcess]) -> Result<()> {
    for process in processes {
        if !process.is_running()? {
            return Err(Error::Process {
                name: process.name.clone(),
                message: format!("exited with status {:?}", process.return_code()?),
            });
        }
    }
    Ok(())
}

fn print_process_logs(processes: &[ManagedProcess]) -> Result<()> {
    for process in processes {
        let output = process.log()?;
        if !output.is_empty() {
            eprintln!("--- {} log ---\n{output}", process.name);
        }
    }
    Ok(())
}
